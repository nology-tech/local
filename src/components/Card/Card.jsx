import { useState } from "react";
import Button from "../Button/Button";
import DetailsPane from "../DetailsPane/DetailsPane";
import "./Card.scss";
import cardCategories from "../../data/favourites/cardCategoryData/cardCategories";

const Card = ({ cardArray = [], primaryButtonOnClick }) => {
  const [activeDetails, setActiveDetails] = useState(null);

  const getDescription = (description, website) => {
    if (!description)
      return `No description, try ${
        website ?? "searching the name and postcode"
      }`;

    if (activeDetails) return description;

    const lastSentenceIndex = description.indexOf(".", 2) + 1;
    const shortenedDescription = description.substring(0, lastSentenceIndex);

    return shortenedDescription;
  };

  const showCards = cardArray.map((card) => {
    const description = getDescription(card?.description, card?.website);

    return (
      <div className="card" key={card._id}>
        <div className="card__header-container">
          <h2
            className={`card__heading card__heading--${
              cardCategories[card.category].categoryName ??
              cardCategories["Other"].categoryName
            }`}
          >
            {card.placename}
          </h2>
          <div className="card__image-container">
            <img
              src={cardCategories[card.category].image}
              alt=""
              className="card__image"
            />
          </div>
        </div>
        <p className="card__text">{description}</p>
        <p className="card__text card__text--adress">{card.address}</p>
        <div className="card__buttons-container">
          <Button
            buttonName="details"
            buttonText="Details"
            onClick={() => setActiveDetails(card)}
          />
          <Button
            buttonName="remove"
            buttonText="Remove"
            onClick={() => primaryButtonOnClick(card._id)}
          />
        </div>
      </div>
    );
  });
  return (
    <>
      {showCards}
      {activeDetails && (
        <DetailsPane
          card={activeDetails}
          onClick={() => setActiveDetails(null)}
        />
      )}
    </>
  );
};

export default Card;
