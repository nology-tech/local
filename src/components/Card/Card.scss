@use "../../../src/styles/variables" as *;
@use "../../styles/mixins" as *;

.card {
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
  border: 2px solid $black;
  border-radius: 1.7rem;
  background-color: $white;
  padding: 1rem;

  .card__header-container {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    flex-grow: 1;
    .card__heading {
      flex-grow: 1;
      font-weight: bold;
      background-color: $green;
      color: $alternate-text-color;
      border-radius: 1.7rem;
      padding: 0.7rem;
      width: 80%;
      margin: auto;
      &--food {
        background-color: $yellow;
        color: $text-color;
      }
      &--culture {
        background-color: $purple;
        color: $alternate-text-color;
      }
    }
    .card__image-container {
      padding: 0.5rem;
      .card__image {
        margin: auto;
      }
    }
  }

  .card__text {
    margin-bottom: auto;
    width: 100%;
    padding: 0 0.7rem;
    flex-grow: 1;
  }
  .card__text--adress {
    font-style: italic;
    font-weight: bold;
    flex-grow: 0;
  }
  .card__buttons-container {
    padding: 0.5rem;
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-around;
    align-items: stretch;
    width: 100%;
  }
}
.card--favourites,
.card--home-list-view {
  margin-bottom: 0.5rem;
}
.card--home {
  margin: 0.5rem;
  height: 10rem;
}
