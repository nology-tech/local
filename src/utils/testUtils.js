import { jest } from "@jest/globals";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import UserContext from "../context/UserContext";
import mockUser from "../data/mockUser";

export const customRender = (
  ui,
  useRouting = true,
  contextOptions = {
    useContext: true,
    value: { user: mockUser, setUser: jest.fn() },
  }
) => {
  // wrap components in routing and context if requested
  const routingUiResult = useRouting && wrapWithRouting(ui);
  const contextUiResult =
    contextOptions.useContext &&
    wrapWithContext(routingUiResult, contextOptions.value);

  // use RTL's render function to return the test component
  return render(contextUiResult);
};

const wrapWithRouting = (ui) => {
  return <Router>{ui}</Router>;
};

const wrapWithContext = (ui, value) => {
  return <UserContext.Provider value={value}>{ui}</UserContext.Provider>;
};
