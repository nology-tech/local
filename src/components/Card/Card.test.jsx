import { customRender } from "../../utils/testUtils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Card from "./Card";
import favArray from "../../data/favourites/favourites";

it("should render the text on the display", () => {
  const { container } = customRender(<Card cardArray={favArray} />);

  expect(container).toMatchSnapshot();
});

it("should check if the click of the details button shows the details pane", () => {
  //load the page
  customRender(<Card cardArray={favArray} />);
  //get the button
  const button = screen.getAllByRole("button");
  //on click of the button
  userEvent.click(button[0]);
  const saveButton = screen.getByText("Save");
  //expect the details pane to appear
  expect(saveButton).toBeInTheDocument();
});
