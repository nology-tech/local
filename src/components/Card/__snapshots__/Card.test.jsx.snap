// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the text on the display 1`] = `
<div>
  <div
    class="card card--undefined"
  >
    <div
      class="card__header-container"
    >
      <h2
        class="card__heading card__heading--food"
      />
      <div
        class="card__image-container"
      >
        <img
          alt=""
          class="card__image"
          src="forkIcon.svg"
        />
      </div>
    </div>
    <p
      class="card__text"
    />
    <p
      class="card__text card__text--adress"
    >
      32 Camden Hill, Eeling, London
    </p>
    <div
      class="card__buttons-container"
    >
      <button
        class="button button--details"
      >
        Details
      </button>
      <button
        class="button button--remove"
      >
        Remove
      </button>
    </div>
  </div>
  <div
    class="card card--undefined"
  >
    <div
      class="card__header-container"
    >
      <h2
        class="card__heading card__heading--other"
      />
      <div
        class="card__image-container"
      >
        <img
          alt=""
          class="card__image"
          src="other.svg"
        />
      </div>
    </div>
    <p
      class="card__text"
    >
      Irish Bar, Great guines on tap.
    </p>
    <p
      class="card__text card__text--adress"
    >
      36 Lamden Hill, Eeling, London
    </p>
    <div
      class="card__buttons-container"
    >
      <button
        class="button button--details"
      >
        Details
      </button>
      <button
        class="button button--remove"
      >
        Remove
      </button>
    </div>
  </div>
  <div
    class="card card--undefined"
  >
    <div
      class="card__header-container"
    >
      <h2
        class="card__heading card__heading--culture"
      />
      <div
        class="card__image-container"
      >
        <img
          alt=""
          class="card__image"
          src="paintIcon.svg"
        />
      </div>
    </div>
    <p
      class="card__text"
    />
    <p
      class="card__text card__text--adress"
    >
      112 Backden Hill, Eeling, London
    </p>
    <div
      class="card__buttons-container"
    >
      <button
        class="button button--details"
      >
        Details
      </button>
      <button
        class="button button--remove"
      >
        Remove
      </button>
    </div>
  </div>
</div>
`;
